services/web/src/components/SearchDropdown.js

```diff
@@ -1,115 +1,113 @@
 import React from 'react';
+import PropTypes from 'prop-types';
 import { Dropdown } from 'semantic-ui-react';
-import { get, omit, flatten, uniqBy } from 'lodash';
+import { omit, uniqBy, debounce } from 'lodash';
 
 export default class SearchDropdown extends React.Component {
-  static defaultProps = {
-    valueField: 'id',
-    textField: 'name',
```

Comment:
should this be removed ?

Reply:
yeah maybe...didn't wanna go too crazy on this component to start

Reply:
you are just hardcoded the fields ? to name and id. i am reading that right.

i think its pretty useful to have the options to be configurable + 
a custom rendering of the text field would also be a win in my book

Reply:
Ah sorry I misunderstood your comment.

First off I think there are certain places where we can reduce complexity by using our own conventions. The `valueField` is definitely one of those... any generic search dropdown should be using IDs those are our bread and butter.

However I see your point about `textField` I was maybe too ambitious about that one. I'm hard-pressed to think of a model we have that doesn't have a name associated with it. My thinking was that anything that can be used inside a search dropdown should have a short field and that field should be called `name`. We could start enforcing this by convention. I didn't communicate this well however (ie. at all) :)

For example yes of course we could have a model that doesn't have a `name` field but instead has a `title` field. However I don't think arbitrarily long titles are going to fit inside a semantic multi Dropdown. Or to rephrase, to be able to use `SearchDropdown` a model has to have a short name that we call `name` by convention.

If there are other real-world examples that you have of an `textField` that isn't `name`, then I'm interested in the use case.

If you just want to have it configurable for the sake of it, I want to start pushing back a BIT on that by having more convention over configuration. I won't die on that hill but for example if we're going to introduce another prop (read: another thing to remember), then what about making it even more flexible, for example:

```jsx
SearchDropdown.propTypes = {
   getText: PropTypes.func
};
SearchDropdown.defaultProps = {
  getText: (item) => item.name,
};
```

could allow something like:

```jsx
<SearchDropdown
   fetchData={this.fetchThings}
   getText={(thing) => `${thing.someOtherNameLikeField} (${thing.type})`}
/>
```

...etc

Reply:
> If there are other real-world examples that you have of an textField that isn't name, then I'm interested in the use case.

Yes, lots. For E-Flux we have several objects that need to be displayed by for example `ocppIdentity` - `serialNumber` in case of devices. For users in Openclear we need a `patient.firstName` and `patient.lastName`. In the case of E-Flux there are so many users that we need to include the email address soon along with the name

Reply:
Also, we have run into issues with `SearchDropdown` for the following things:

- Being able to search on a field other than `name` or the field that's used to display the data. For example searching by keyword and displaying name.
- Having the actual selected object back on the onChange can be super helpful in some cases (e.g. location selector on Openclear)

Reply:
awesome ok... I was asking because I really am interested in the use cases :)

ok lemme take those and run with it...

btw I think `onChange` should ALWAYS speak actual objects I've spent cycles trying to undo tapdancing between objects and their ids it's silly... have some ideas around this for later

Reply:
OK this is great... rather than go back and forth on this into the future I've taken those use cases and turned them into test cases so we'll be sure that this component can handle what we need ... we can add more cases as needed. Going to start doing this for more bedrock components.

For `SearchDropdown` specifically I've done a few things:

- keeping the hardcoded `id` field for `value`... our API speaks ids
- renamed `fetchData` to `onDataNeeded` to align with `SearchProvider`...also I don't like the name `fetch`... that's an implementation detail the provider doesn't need to be aware of
- added `getItemLabel` prop with a default of `(item) => item.name`

This should handle @dominiek's use cases above:

```jsx
<SearchProvider
  onDataNeeded={this.fetchUsers}
  getItemLabel={(user) => {
    return `${user.firstName} ${user.lastName} (${user.email})`;
  }}
/>
```

```jsx
<SearchProvider
  onDataNeeded={this.fetchWonkyObjects}
  getItemLabel={({ ocppIdentity, serialNumber }) => {
    return `${ocppIdentity} - ${serialNumber}`;
  }}
/>
```

> Being able to search on a field other than name

`onDataNeeded` will now receive a single string argument that is the search query. It can query on any field it wants to

> Having the actual selected object back on the onChange can be super helpful

Turns out I already made this change. `SearchDropdown` is now essentially a generic form field that handles type `ObjectId`s ... it consumes populated objects as `value` and returns them as `value` in `onChange`. We can avoid a lot of react wonkiness this way with the "is it an object or is it a string" song and dance... it always deals with objects not ids.

Reply:
could we make it `getItem(item)` or perhaps `getOption(item)` and let that user specify the both { value, text }.
The id might not be id. More rare, but I have that in both e-flux and acrylic.

Reply:
> The id might not be id. More rare, but I have that in both e-flux and acrylic.

The ID is considered the de-facto unique identifier for objects that come out of our APIs. If there's a real world use case to not speak IDs then let's have it and I'll add a test case. I will also add a `getOption` or `getOptionValue` function but I will keep `getOptionLabel` as it's by far the bigger use case. What I don't want is to make increasingly obtuse APIs for the sake of vanishingly uncommon use cases.

Reply:
also sorry that should be `getOptionLabel` above not `getItemLabel` ... trying to crack down on these mystery meat terms like `item`

Reply:
ok resolution here a custom `id` field we've agreed is a very uncommon use case. Want to see a fresh use case for this one then I can easily add it back in if needed